view: Refactor view->output usage

jlindgren90 opened the pull request:
Needs more thorough testing but I'm looking for initial feedback on the approach.

This comes with two (intentional) behavior changes:
- Maximized/tiled views now try harder to remain on the same output if the output layout changes
- Initially maximized xdg-shell views now appear on the cursor output (resolves #767)

`view->output` is now used as follows (copy & pasting from `view.h`):
```
/*
 * The output that the view is displayed on. Specifically:
 *
 *  - For floating views, this is the output nearest to the
 *    center of the view. It is computed automatically when the
 *    view is moved or the output layout changes.
 *
 *  - For fullscreen/maximized/tiled views, this is the output
 *    used to compute the view's geometry. The view remains on
 *    the same output unless it is disabled or disconnected.
 *
 * Many view functions (e.g. view_center(), view_fullscreen(),
 * view_maximize(), etc.) allow specifying a particular output
 * by setting view->output explicitly before calling them.
 */
```

Consolatis commented on src/xdg.c:
```diff
@@ -175,17 +175,33 @@ handle_request_maximize(struct wl_listener *listener, void *data)
+		} else {
+			wlr_log(WLR_ERROR,
+				"invalid output in fullscreen request");
+		}
```
I am not sure right now but I think clients can ask for fullscreen without supplying an output.
In the case of `!view->been_mapped` that should likely be the cursor output, otherwise the output the view is mainly on.

jlindgren90 replied:
That should be handled all right; view_set_fullscreen() is called regardless and will use the current view->output in that case.

Consolatis commented on src/xdg.c:
```diff
@@ -175,17 +175,33 @@ handle_request_maximize(struct wl_listener *listener, void *data)
+		struct output *output = output_from_wlr_output(view->server,
+			requested->fullscreen_output);
+		if (output_is_usable(output)) {
+			view->output = output;
```
This directly modifies the output for the view, what do you think about something like `view_set_output(view, output)` instead?

jlindgren90 replied:
`view_set_output(view, output)` might be nicer.

Inside `view_set_output`, we could do `assert(!view->fullscreen)` (or maybe just return) since modifying `view->output` for a fullscreen view isn't allowed.

Consolatis commented on src/view.c:
```diff
@@ -342,15 +354,11 @@ view_apply_region_geometry(struct view *view)
 	assert(view);
 	assert(view->tiled_region || view->tiled_region_evacuate);
+	struct output *output = view->output;
+	assert(output_is_usable(output));
```
Couldn't this assert trigger when there are not outputs left?

jlindgren90 replied:
The calling paths should prevent that now, see the new checks in `view_apply_special_geometry()` and `view_snap_to_region()`.

Consolatis commented:
In theory I like this approach. I am a bit worried how this reacts in the case of no outputs being available at all.
Including for fullscreen / maximize / tiling requests (from the client or the user via keybind), newly mapped windows and simliar.

jlindgren90 commented:
> In theory I like this approach. I am a bit worried how this reacts in the case of no outputs being available at all. Including for fullscreen / maximize / tiling requests (from the client or the user via keybind), newly mapped windows and simlar.
> 
> Edit: I guess if we prevent any `apply_xxx_geometry()` to be called in those cases it would be fine.

There is a top-level check for a usable output in `view_apply_special_geometry()` which is where (with a few exceptions) most of the other `apply` functions are called from. I think I checked the "few exceptions" too and made sure we have a usable output in those cases (hence the new checks to `view_move_to_edge` + `view_snap_to_edge` + `view_snap_to_region`) but definitely would appreciate a second pair of eyes (and testing!) to make sure I didn't miss any.

johanmalm commented:
Other than Consolatis's comments, it looks fine to me. I like the approach - seems to simplify things :smile:

jlindgren90 commented:
Rebased to fix merge conflicts. I still need to add `view_set_output()`.

jlindgren90 commented:
I added `view_set_output()` (new commit on top; probably not much value in rewriting the previous ones). I think this should be all set now, from my point-of-view.

Consolatis commented:
I'll try to test this later today.

Consolatis commented on include/view.h:
```diff
@@ -129,18 +146,13 @@ void view_move_resize(struct view *view, struct wlr_box geo);
-/* output is optional, defaults to current nearest output */
-void view_center(struct view *view, struct output *output,
-	const struct wlr_box *ref);
+void view_center(struct view *view, const struct wlr_box *ref);
```
When reading the function definition it doesn't immediately become clear to me what `*ref` means.
So maybe we could add either a comment above explaining it or rename the argument. Maybe something like `container` or `parent` would make it more clear?

jlindgren90 replied:
I'll add a comment, seems worth explaining.

Consolatis commented on src/view.c:
```diff
@@ -102,6 +102,21 @@ view_get_edge_snap_box(struct view *view, struct output *output,
+ * At present, a view can only 'enter' one output at a time, although the view
+ * may span multiple outputs. Ideally we would handle multiple outputs, but
+ * this method is the simplest form of what we want.
+ */
```
Is this comment still valid, e.g. do we actually need to know if a view is on multiple outputs?

Consolatis commented on src/view.c:
```diff
@@ -102,6 +102,21 @@ view_get_edge_snap_box(struct view *view, struct output *output,
 	return dst;
 }
 
+/*
+ * At present, a view can only 'enter' one output at a time, although the view
+ * may span multiple outputs. Ideally we would handle multiple outputs, but
+ * this method is the simplest form of what we want.
+ */
+static void
+view_discover_output(struct view *view)
```
Is this comment still valid, e.g. do we actually need to know if a view is on multiple outputs?

jlindgren90 replied:
I'm good with removing this comment. It seems less relevant now.

Consolatis commented on src/view.c:
```diff
@@ -156,7 +183,14 @@ view_moved(struct view *view)
 	wlr_scene_node_set_position(&view->scene_tree->node,
 		view->current.x, view->current.y);
-	view_discover_output(view);
+	/*
+	 * Only floating views change output when moved. Non-floating
+	 * views (maximized/tiled/fullscreen) are tied to a particular
+	 * output when they enter that state.
+	 */
+	if (view_is_floating(view)) {
+		view_discover_output(view);
+	}
```
I didn't investigate this in detail yet, just wanted to point out that this *might* be an issue for interactive move from tiled state because we delay `view_set_untiled(view)`, I am not quite sure right now if `view_moved()` is called before the tiling state is reset.

johanmalm replied:
I've tested to/from tiled state with mouse/keyboard. Works fine.

Consolatis commented:
I had a quick test nested and views in various "special" states migrated fine when closing one of the virtual outputs.
Also tried opening new windows with no outputs left and apart from some error logs everything seems fine. This whole PR looks like a really nice improvement and we finally get rid of those `view->current.x/y` hacks to define the initial output.

johanmalm commented:
LGTM.

I note Consolatis's comments from yesterday, but otherwise seems ready to merge.

jlindgren90 commented:
I added a comment for `view_center` and deleted the obsolete one for `view_discover_output`. I haven't had a chance yet to test moving a view from the tiled state (currently spending time debugging an amdgpu driver regression).
